### DISCUSSION TITLE
Custom PageContext fields not available in onBeforeRender hook and other router functions when on Client Routing?

### DISCUSSION THREAD

**Discussion by apappas1129** [vikejs/vike#1039] [Discussion]
I just found out that the custom fields I injected on `pageContextInit` like CASL ablilities, and user is not available. the `pageContext` in `onBeforeRender` is straight up only the `PageContextBuiltIn` ?

server
```typescript
const pageContextInit: Partial<PageContextServer> = {
    urlOriginal: req.originalUrl,
    user: req.user,
    ability: req.ability,
  };
...
const pageContext = await renderPage(pageContextInit);
``` 

*.page.ts (comes with a *.page.client.vue if that matters)
```typescript
export { onBeforeRender };

async function onBeforeRender(pageContext: PageContextBuiltIn) {
// pageContext.user undefined
// pageContext.ability undefined
``` 

I need it for page redirection logic like for example:

```typescript
async function onBeforeRender(pageContext: PageContextBuiltIn) {
  const redirect = { pageContext: { redirectTo: '/subjects' } };
  if (!pageContext.headers?.cookie || !pageContext.ability) return redirect;

  const unpackedRules = unpackRules(pageContext.ability);
  const ability = createAbility(unpackedRules);
  
  if (!ability.can('update', subjectHelper('content', { authorId: pageContext.user!.id }))) {
    return redirect;
  }
```

(using [client routing redirect](https://vite-plugin-ssr.com/page-redirection#client-side))

Also true for `guard`. Maybe this is meant to be done somewhere else?

**Comment by brillout** [comment]
`pageContextInit` should be included. If it isn't then it's a bug.

There is actually a bug fix about this in the `dev` branch which will should be released soon. (ETA in the next couple of days.)

In the meantime, can you define `onBeforeRender()` in `.page.server.js`?
